prometheus-net/Histogram.cs

```diff
@@ -48,55 +47,35 @@ internal Histogram(string name, string help, string[] labelNames, double[] bucke
 
         public class Child : Advanced.Child, IHistogram
         {
-            private double _sum = 0;
-            private ulong _count = 0;
-            private ulong[] _bucketCounts;
-            private readonly object _lock = new object();
-
-            //re-using WireFormat.Histogram to avoid excessive array allocations 
-            //TODO: this will break thread-safety if the very same histogram instance is registered on multiple CollectorRegistries
-            //TODO: we should investigate if HdrHistogram can be a good candidate to use here internally - that's heavily optimized
-
-            private Advanced.DataContracts.Histogram _wireMetric;
-            private double[] _buckets;
+            private ThreadSafeDouble _sum = new ThreadSafeDouble(0.0D);
+            private ThreadSafeLong[] _bucketCounts;
+            private double[] _upperBounds;
             
             internal override void Init(ICollector parent, LabelValues labelValues)
             {
                 base.Init(parent, labelValues);
 
-                _buckets = ((Histogram) parent)._buckets;
-
-                if (_buckets == null)
-                {
-                    return;
-                }
-
-                _bucketCounts = new ulong[_buckets.Length];
-                _wireMetric = new Advanced.DataContracts.Histogram();
-                for (int i = 0; i < _bucketCounts.Length; i++)
-                {
-                    _wireMetric.bucket.Add(new Bucket
-                    {
-                        upper_bound = _buckets[i]
-                    });
-                }
+                _upperBounds = ((Histogram)parent)._buckets;
+                _bucketCounts = new ThreadSafeLong[_upperBounds.Length];
             }
 
             protected override void Populate(Metric metric)
             {
-                lock (_lock)
+                var wireMetric = new Advanced.DataContracts.Histogram();
+                wireMetric.sample_count = 0L;
+
+                for (var i = 0; i < _bucketCounts.Length; i++)
                 {
-                    _wireMetric.bucket[0].cumulative_count = _bucketCounts[0];
-                    for (int i = 1; i < _bucketCounts.Length; i++)
+                    wireMetric.sample_count += (ulong)_bucketCounts[i].Value;
+                    wireMetric.bucket.Add(new Bucket
                     {
-                        _wireMetric.bucket[i].cumulative_count = _bucketCounts[i] + _wireMetric.bucket[i - 1].cumulative_count;
-                    }
-                    _wireMetric.sample_count = _count;
-                    _wireMetric.sample_sum = _sum;
+                        upper_bound = _upperBounds[i],
+                        cumulative_count = wireMetric.sample_count
```

Comment:
The accumulated count for the current index is the same as the cumulative count for the current bucket. This avoids the need to lookup into a previous bucket which can cause differences in counts throughout due threading when we remove the lock.  By reading only one value per bucket we allow the total count and the counts of each bucket to remain internally consistent for a collection without using locks.